models/image.py

```diff
@@ -1242,6 +1077,244 @@ def load(self):
+        psf_provs = []
+        wcs_provs = []
+        zp_provs = []
+
+        for im in self.upstream_images:
+            image_provs.append(im.provenance)
+            if im.sources is not None:
+                sources_provs.append(im.sources.provenance)
+            if im.psf is not None:
+                psf_provs.append(im.psf.provenance)
+            if im.wcs is not None:
+                wcs_provs.append(im.wcs.provenance)
+            if im.zp is not None:
+                zp_provs.append(im.zp.provenance)
+
+        if len(set([p.id for p in image_provs])) > 1:
+            raise ValueError("Cannot combine images with different provenance IDs.")
+        if len(set([p.id for p in sources_provs])) > 1:
+            raise ValueError("Cannot combine sources with different provenance IDs.")
+        if len(set([p.id for p in psf_provs])) > 1:
+            raise ValueError("Cannot combine PSFs with different provenance IDs.")
+        if len(set([p.id for p in wcs_provs])) > 1:
+            raise ValueError("Cannot combine WCSs with different provenance IDs.")
+        if len(set([p.id for p in zp_provs])) > 1:
+            raise ValueError("Cannot combine ZPs with different provenance IDs.")
+
+        return image_provs + sources_provs + psf_provs + wcs_provs + zp_provs
+
+    def load_upstream_products(self, session=None):
+        """Make sure each upstream image has its related products loaded.
+
+        This only works after all the images and products are committed to the database,
+        with provenances consistent with what is saved in this Image's provenance
+        and its own upstream_ids.
+        """
+        prov_ids = self.provenance.upstream_ids
+        # check to make sure there is any need to load
+        need_to_load = False
+        for im in self.upstream_images:
+            if im.sources is None or im.sources.provenance_id not in prov_ids:
+                need_to_load = True
+                break
+            if im.psf is None or im.psf.provenance_id not in prov_ids:
+                need_to_load = True
+                break
+            if im.wcs is None or im.wcs.provenance_id not in prov_ids:
+                need_to_load = True
+                break
+            if im.zp is None or im.zp.provenance_id not in prov_ids:
+                need_to_load = True
+                break
+
+        if not need_to_load:
+            return
+
+        from models.source_list import SourceList
+        from models.psf import PSF
+        from models.world_coordinates import WorldCoordinates
+        from models.zero_point import ZeroPoint
+
+        with SmartSession(session) as session:
+            im_ids = [im.id for im in self.upstream_images]
+
+            # get all the products for all images at once
+            sources_result = session.scalars(
+                sa.select(SourceList).where(
+                    SourceList.image_id.in_(im_ids),
+                    SourceList.provenance_id.in_(prov_ids),
```

Comment:
This is going to have trouble again....  It's possible that the various provenances that went into the reference image for a subtraction will not match the current provenances, as I note elsewhere.  That means the source list won't have the same provenance (because of upstreams), even if it was run using the same source list code.

In fact, this will have trouble in a subtraction even if nothing in the code has changed.  The upstream ref is probably a coadd, but the new could well be a single image, so their provenances won't match (and thus the provenances of things like source lists that come out of them won't match).